presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -296,16 +305,65 @@ public ConnectorSplitSource getSplits(
                 splitSchedulingContext.schedulerUsesHostAddresses(),
                 layout.isPartialAggregationsPushedDown());
 
-        HiveSplitSource splitSource = computeSplitSource(splitSchedulingContext, table, session, hiveSplitLoader);
+        HiveSplitSource splitSource = computeSplitSource(splitSchedulingContext, table, session, hiveSplitLoader, ratio);
         hiveSplitLoader.start(splitSource);
 
         return splitSource;
     }
 
+    private double getSplitScanRatio(
+            ConnectorSession session,
+            SchemaTableName tableName,
+            HiveTableLayoutHandle layout,
+            TransactionalMetadata metadata,
+            List<HivePartition> partitions)
+    {
+        double ratio = 1.0;
+        if (!isDynamicSplitSizesEnabled(session)) {
+            return ratio;
+        }
+        HiveTableHandle hiveTableHandle = new HiveTableHandle(tableName.getSchemaName(), tableName.getTableName());
+
+        Set<HiveColumnHandle> readColumnHandles = new HashSet<>();
+        if (layout.getRequestedColumns().isPresent()) {
+            readColumnHandles.addAll(layout.getRequestedColumns().get());
+        }
+        readColumnHandles.addAll(layout.getPredicateColumns().values());
+        readColumnHandles.removeIf(handle -> handle.isHidden() || handle.isPartitionKey());
+
+        Map<String, Type> readColumnTypes = readColumnHandles.stream().collect(toImmutableMap(HiveColumnHandle::getName, handle -> metadata.getColumnMetadata(session, hiveTableHandle, handle).getType()));
+        Map<String, ColumnHandle> readColumns = readColumnHandles.stream().collect(toImmutableMap(HiveColumnHandle::getName, identity()));
+
+        TableStatistics tableStatistics = metadata.getHiveStatisticsProvider().getTableStatistics(session, tableName, readColumns, readColumnTypes, partitions);
+        double totalSize = tableStatistics.getTotalSize().getValue();
+        double requiredSize = 0;
+        double rowCount = tableStatistics.getRowCount().getValue();
+        for (Map.Entry<ColumnHandle, ColumnStatistics> entry : tableStatistics.getColumnStatistics().entrySet()) {
+            double value = entry.getValue().getDataSize().getValue();
+            // We do not compute column stats for fixed width types, so count them manually.
+            if (!isFinite(value) && isFinite(rowCount)) {
+                HiveColumnHandle columnHandle = (HiveColumnHandle) entry.getKey();
+                Type type = metadata.getColumnMetadata(session, hiveTableHandle, columnHandle).getType();
+                if (type instanceof FixedWidthType) {
+                    int size = ((FixedWidthType) type).getFixedSize();
+                    value = size * rowCount;
+                }
+            }
+            requiredSize += value;
+            readColumnHandles.remove(entry.getKey());
```

Comment:
why are we removing a value from readColumnHandles? It doesn't look like we use it again.